src/many-ledger/src/storage.rs

```diff
@@ -333,6 +538,398 @@ impl LedgerStorage {
+                .unwrap_or(MULTISIG_DEFAULT_THRESHOLD),
+        };
+        let timeout_in_secs = match arg.timeout_in_secs {
+            Some(t) if is_owner => t,
+            Some(_) => return Err(account::errors::user_needs_role("owner")),
+            _ => multisig_f
+                .arg
+                .timeout_in_secs
+                .unwrap_or(MULTISIG_DEFAULT_TIMEOUT_IN_SECS),
+        }
+        .min(MULTISIG_MAXIMUM_TIMEOUT_IN_SECS);
+        let execute_automatically = match arg.execute_automatically {
+            Some(e) if is_owner => e,
+            Some(_) => return Err(account::errors::user_needs_role("owner")),
+            _ => multisig_f
+                .arg
+                .execute_automatically
+                .unwrap_or(MULTISIG_DEFAULT_EXECUTE_AUTOMATICALLY),
+        };
+        let time = self.current_time.unwrap_or_else(SystemTime::now);
+
+        // Calculate the approver list, set their approvals to false, except for
+        // the sender.
+        let mut approvers = BTreeMap::new();
+        approvers.insert(*sender, multisig::ApproverInfo { approved: true });
+        for (id, roles) in account.roles() {
+            if roles.contains(MULTISIG_ROLE_CAN_APPROVE) || roles.contains(MULTISIG_ROLE_CAN_SUBMIT)
+            {
+                approvers.entry(*id).or_default();
+            }
+        }
+
+        let timeout = Timestamp::from(time + Duration::from_secs(timeout_in_secs));
+        let storage = TransactionStorage {
+            account: account_id,
+            info: multisig::InfoReturn {
+                memo: arg.memo.clone(),
+                transaction: arg.transaction.as_ref().clone(),
+                submitter: *sender,
+                approvers,
+                threshold,
+                execute_automatically,
+                timeout,
+                data: arg.data.clone(),
+            },
+        };
+
+        self.commit_multisig_transaction(tx_id.0.as_slice(), &storage)?;
+        self.add_transaction(TransactionInfo::AccountMultisigSubmit {
+            submitter: *sender,
+            account: account_id,
+            memo: arg.memo,
+            transaction: Box::new(*arg.transaction),
+            token: Some(tx_id.0.clone()),
+            threshold,
+            timeout,
+            execute_automatically,
+            data: arg.data,
+        });
+
+        Ok(tx_id.0.to_vec())
+    }
+
+    pub fn get_multisig_info(&self, tx_id: &[u8]) -> Result<TransactionStorage, ManyError> {
+        let storage_bytes = self
+            .persistent_store
+            .get(&key_for_multisig_transaction(tx_id))
+            .unwrap_or(None)
+            .ok_or_else(multisig::errors::transaction_cannot_be_found)?;
+        minicbor::decode::<TransactionStorage>(&storage_bytes)
+            .map_err(|e| ManyError::deserialization_error(e.to_string()))
+    }
+
+    pub fn approve_multisig(&mut self, sender: &Identity, tx_id: &[u8]) -> Result<bool, ManyError> {
```

Comment:
nit:

We need to test corner-cases such as

- A multisig transaction has been submitted. The list of approvers is X
- We remove the `canMultisigApprove` role from one of the user in X

Here, we never check that the sender has the `canMultisigApprove` role.

We should at least document the intended behavior.

Reply:
Okay I'll refactor here to be more in line with the spec. I think the current behaviour is unexpected yes.